File: NCoreUtils.AspNetCore.Rest/DefaultRestCreate.fs
namespace NCoreUtils.AspNetCore.Rest

open System.Runtime.CompilerServices
open Microsoft.Extensions.Logging
open NCoreUtils.AspNetCore
open NCoreUtils.Data
open NCoreUtils.Linq
open NCoreUtils.Logging

/// <summary>
/// Provides default implementation for REST CREATE method.
/// </summary>
/// <typeparam name="a">Type of the target object.</typeparam>
/// <typeparam name="id">Type of the Id property of the target object.</typeparam>
type DefaultRestCreate<'a, 'id when 'a :> IHasId<'id> and 'id : equality> =
  inherit DefaultTransactedMethod<'a, 'id>
  val private logger : ILogger

  /// <summary>
  /// Initializes new instance from the specified parameters.
  /// </summary>
  /// <param name="repository">Repository to use.</param>
  /// <param name="loggerFactory">Logger factory.</param>
  [<MethodImpl(MethodImplOptions.AggressiveInlining)>]
  new (repository, loggerFactory : ILoggerFactory) =
    { inherit DefaultTransactedMethod<'a, 'id> (repository)
      logger     = loggerFactory.CreateLogger "NCoreUtils.AspNetCore.Rest.DefaultRestCreate" }

  abstract AsyncInvoke : data:'a -> Async<'a>

  /// <summary>
  /// Performes REST CREATE action for the predefined type.
  /// </summary>
  /// <param name="data">Object to insert into dataset.</param>
  /// <returns>
  /// Object returned by dataset after insert operation. Depending on the repository implementation some of the values
  /// of the returned object may differ from the input.
  /// </returns>
  default this.AsyncInvoke (data : 'a) = async {
    // check if already exists
    if data.HasValidId () then
      let! exists =
        let id = data.Id
        this.Repository.Items |> Q.asyncExists (fun item -> item.Id = id)
      do
        match exists with
        | true ->
          debugf this.logger "Entity of type %s with key = %A already exists (data-create)." typeof<'a>.FullName data.Id
          ConflictException "Entity already exists." |> raise
        | _ -> () // BadRequestException "New entities must not include id." |> raise
    // persist entity
    let! item = this.Repository.AsyncPersist data
    debugf this.logger "Entity of type %s has been created with key = %A (data-create)." typeof<'a>.FullName item.Id
    return item }
  interface IRestCreate<'a, 'id> with
    member this.AsyncInvoke data = this.AsyncInvoke data
  interface IBoxedInvoke<'a, 'a> with
    member this.Instance = box this
    member this.AsyncInvoke data = this.AsyncInvoke data

File: NCoreUtils.AspNetCore.Rest/Rest/DefaultRestCreate.cs
using System;
using System.Diagnostics.CodeAnalysis;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using NCoreUtils.Data;
using NCoreUtils.Linq;

namespace NCoreUtils.AspNetCore.Rest;

/// <summary>
/// Provides default implementation for REST CREATE method.
/// </summary>
/// <typeparam name="TData">Type of the target object.</typeparam>
/// <typeparam name="TId">Type of the Id property of the target object.</typeparam>
/// <remarks>
/// Initializes new instance from the specified parameters.
/// </remarks>
/// <param name="repository">Repository to use.</param>
/// <param name="logger">Logger to use.</param>
public class DefaultRestCreate<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicProperties)] TData, TId>(
    IDataRepository<TData, TId> repository,
    ILogger<DefaultRestCreate<TData, TId>> logger)
    : DefaultTransactedMethod<TData, TId>(repository)
    , IRestCreate<TData, TId>
    , IBoxedInvoke<IRestCreateContext<TData, TId>, TData>
    where TData : IHasId<TId>
{
    protected ILogger Logger { get; } = logger ?? throw new ArgumentNullException(nameof(logger));

    object IBoxedInvoke.Instance => this;

    protected virtual bool HasValidId(TData data)
        => data.HasValidId();

    /// <summary>
    /// Performes REST CREATE action for the predefined type.
    /// </summary>
    /// <param name="context">Invocation context.</param>
    /// <param name="cancellationToken">Cancellation token.</param>
    /// <returns>
    /// Object returned by dataset after insert operation. Depending on the repository implementation some of the values
    /// of the returned object may differ from the input.
    /// </returns>
#if NET7_0
    [UnconditionalSuppressMessage("AOT", "IL3050", Justification = "Handled by query provider.")]
    [UnconditionalSuppressMessage("Trim", "IL2026", Justification = "Handled by query provider.")]
#endif
    public virtual async ValueTask<TData> InvokeAsync(IRestCreateContext<TData, TId> context, CancellationToken cancellationToken)
    {
        var data = context.ThrowIfNull().Data;
        if (data.HasValidId())
        {
            // check if already exists
            if (await Repository.Items.AnyAsync(context.CreateIdEqualsPredicate(data.Id), cancellationToken))
            {
                Logger.LogRestEntityAlreadyExists(typeof(TData), data.Id);
                throw new ConflictException("Entity already exists.");
            }
        }
        // persist entity
        var result = await Repository.PersistAsync(data, cancellationToken);
        Logger.LogRestEntityCreatedSuccessfully(typeof(TData), result.Id);
        return result;
    }
}